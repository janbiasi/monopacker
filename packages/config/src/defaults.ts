import type { Config, PackConfig, ResolverConfig } from './types';

export const requiredFilesToCopy = ['!package.json', '!node_modules'];

/**
 * Files which should get copied
 */
export const defaultFilesToCopy = [
	'**',
	...requiredFilesToCopy,
	'!.editorconfig',
	'!.gitignore',
	'!.nvmrc',
	'!.node-version',
	'!.npmrc',
	'!**.md',
	'!tsconfig*.json',
];

/**
 * Patterns which should be ignored in every glob
 */
export const defaultIgnorePatterns = [
	'**/node_modules',
	'**/bower_components',
	'**/vendor',
	'**/dist',
	'**/build',
	'**/lib',
	'**/bin',
	'**/test',
	'**/tests',
	'**/__tests__',
	'**/__mocks__',
	'**/__snapshots__',
	'**/__test__',
];

/**
 * Patch default config for a pack configuration
 * @param {PackConfig} config
 * @returns
 */
function patchPackDefaults(config: PackConfig): Required<PackConfig> {
	return {
		source: config.source,
		copy: config.copy ? [...config.copy, ...requiredFilesToCopy] : defaultFilesToCopy,
	};
}

function patchResolverDefaults(config: ResolverConfig): Required<ResolverConfig> {
	return {
		ignore: config.ignore || defaultIgnorePatterns,
		deep: config.deep || 7,
		followSymbolicLinks: config.followSymbolicLinks || true,
		includePeers: config.includePeers || true,
		externals: config.externals || [],
	};
}

export function patchConfigDefaults(config: Config): Required<Config> {
	return {
		rootDir: config.rootDir || './',
		cache: config.cache === false ? false : true,
		internals: config.internals || [],
		packageManager: config.packageManager || 'npm',
		packs: config.packs.map(patchPackDefaults),
		resolver: patchResolverDefaults(config.resolver || {}),
	};
}
