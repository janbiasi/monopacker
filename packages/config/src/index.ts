export { Config, ResolverConfig, PackConfig, PackageManager } from './types';
export { getConfig } from './resolve';
export { validateConfig } from './validator';
export { defaultFilesToCopy, defaultIgnorePatterns } from './defaults';
