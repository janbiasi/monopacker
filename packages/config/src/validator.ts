import type { Config } from './types';

/**
 * Basic validation for the monopacker configuration, it will throw an error in the following cases:
 * - If the configuration is not an object
 * - If the configuration is an object but has no required `packs` property
 * - If the `packs` property is not an array
 * - If the `packs` property is an array but has no required elements
 * @param {Config}Â config The configuration to validate
 */
export const validateConfig = (config: Partial<Config>): void => {
	if (typeof config !== 'object') {
		throw new Error('Monopacker configuration must be an object');
	}

	if (!config.packs) {
		throw new Error('Monopacker configuration must contain a property "packs"');
	}

	if (!Array.isArray(config.packs)) {
		throw new Error(`Monopacker packs must be an array, got ${typeof config.packs} instead`);
	}

	if (config.packs.length === 0) {
		throw new Error('Monopacker packs setting must contain at least one entry');
	}
};
