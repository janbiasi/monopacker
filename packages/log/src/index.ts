export * from './default';
export { LogLevel, LoggableLevel, loggableLevels } from './level';
export { createLogger, getLogStack } from './logger';
export { Logger } from './types';
export { Namespace, namespaces } from './namespace';
