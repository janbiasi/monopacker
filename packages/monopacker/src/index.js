"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
// expose all types
__export(require("./types"));
// expose adapters and blueprint for custom ones
var adapter_1 = require("./adapter");
exports.Adapter = adapter_1.Adapter;
exports.AdapterLerna = adapter_1.AdapterLerna;
// expose main packer
var Packer_1 = require("./Packer");
exports.Packer = Packer_1.Packer;
// expose taper
var Taper_1 = require("./Taper");
exports.Taper = Taper_1.Taper;
// expose helpers
var debug_hooks_1 = require("./helper/debug-hooks");
exports.useDebugHooks = debug_hooks_1.useDebugHooks;
//# sourceMappingURL=index.js.map