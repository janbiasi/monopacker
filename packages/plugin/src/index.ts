export const test = 10;
