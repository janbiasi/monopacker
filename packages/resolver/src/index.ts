export { Graph, GraphResolutionEntry } from './types/graph';
export { PackageJson, PackageDependencyMap, PackagePeerDependenciesMeta } from './types/package';
export { TargetGraph } from './types/target-graph';

export { getPackageGraph, getPackageResolutions } from './graph';
export { parsePackage, parsePackages } from './package';
export { resolveTargetGraph, resolveTargetPackageName } from './resolve';
export { searchPackages, searchFiles } from './search';
