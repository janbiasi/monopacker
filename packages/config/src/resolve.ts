import type { Config } from './types';
import { cosmiconfig } from 'cosmiconfig';
import { validateConfig } from './validator';
import { patchConfigDefaults } from './defaults';
import { logger } from './logger';

const explorer = cosmiconfig('monopacker');

/**
 * Returns the local configuration object by using the common cosmiconfig explorer.
 * The method throws an error if no configuration could be found.
 * It will search in the following places:
 * - `package.json`, property `"monopacker"`
 * - `.monopackerrc` (JSON)
 * - `.monopackerrc.json`
 * - `.monopackerrc.y(a)ml`
 * - `.monopackerrc.(c)js`
 * - `monopacker.config.(c)js`
 * @returns {[Required<Config>, string]} A tuple with the configuration object and the path
 */
export const getConfig = async (rootDir: string): Promise<[Required<Config>, string]> => {
	try {
		logger.debug(`Searching for configuration file in ${rootDir}`);
		const result = await explorer.search(rootDir);

		if (result && result.isEmpty) {
			throw new Error(`Monopacker configuration is empty`)
		}

		if (result && result.config) {
			logger.debug(`Picking up configuration from ${result.filepath}`);
			validateConfig(result.config);

			return [patchConfigDefaults(result.config), result.filepath];
		}


		throw new Error(`No monopacker configuration found`);
	} catch (err) {
		const nativeErrorMessage = err instanceof Error ? err.message : `${err || 'unknown'}`;
		throw new Error(`Failed resolving monopacker configuration in ${rootDir}: ${nativeErrorMessage}`);
	}
};
