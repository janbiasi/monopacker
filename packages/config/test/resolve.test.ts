import { resolve } from 'path';
import { getConfig } from '../src/resolve';

const FIXTURES_DIR = resolve(__dirname, 'fixtures');

const getFixtureRootDir = (name: string) => resolve(FIXTURES_DIR, name);

describe('resolve', () => {
	it('should resolve a dedicated valid config correctly', async () => {
		const [config] = await getConfig(getFixtureRootDir('dedicated-file-valid'));
		expect(config).toBeTruthy();
		expect(config).toMatchSnapshot();
	});

	it('should resolve a package.json config correctly', async () => {
		const [config] = await getConfig(getFixtureRootDir('package-valid'));
		expect(config).toBeTruthy();
		expect(config).toMatchSnapshot();
	});
});
