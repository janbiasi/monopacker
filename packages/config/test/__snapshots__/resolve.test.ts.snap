// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolve should resolve a dedicated valid config correctly 1`] = `
Object {
  "cache": true,
  "internals": Array [],
  "packageManager": "npm",
  "packs": Array [
    Object {
      "copy": Array [
        "**",
        "!package.json",
        "!node_modules",
        "!.editorconfig",
        "!.gitignore",
        "!.nvmrc",
        "!.node-version",
        "!.npmrc",
        "!**.md",
        "!tsconfig*.json",
      ],
      "source": "my-app",
    },
  ],
  "resolver": Object {
    "deep": 7,
    "externals": Array [],
    "followSymbolicLinks": true,
    "ignore": Array [
      "**/node_modules",
      "**/bower_components",
      "**/vendor",
      "**/dist",
      "**/build",
      "**/lib",
      "**/bin",
      "**/test",
      "**/tests",
      "**/__tests__",
      "**/__mocks__",
      "**/__snapshots__",
      "**/__test__",
    ],
    "includePeers": true,
  },
  "rootDir": "./",
}
`;

exports[`resolve should resolve a package.json config correctly 1`] = `
Object {
  "cache": true,
  "internals": Array [],
  "packageManager": "npm",
  "packs": Array [
    Object {
      "copy": Array [
        "**",
        "!package.json",
        "!node_modules",
        "!.editorconfig",
        "!.gitignore",
        "!.nvmrc",
        "!.node-version",
        "!.npmrc",
        "!**.md",
        "!tsconfig*.json",
      ],
      "source": "apps/my-app",
    },
  ],
  "resolver": Object {
    "deep": 7,
    "externals": Array [],
    "followSymbolicLinks": true,
    "ignore": Array [
      "**/node_modules",
      "**/bower_components",
      "**/vendor",
      "**/dist",
      "**/build",
      "**/lib",
      "**/bin",
      "**/test",
      "**/tests",
      "**/__tests__",
      "**/__mocks__",
      "**/__snapshots__",
      "**/__test__",
    ],
    "includePeers": true,
  },
  "rootDir": "./",
}
`;
